devdoc/sm_requirements.md

```diff
@@ -198,6 +201,23 @@ MOCKABLE_FUNCTION(, int, sm_close_begin, SM_HANDLE, sm);
 
 **SRS_SM_02_071: [** If there are any failures then `sm_close_begin` shall fail and return `SM_ERROR`. **]**
 
+### sm_close_begin_with_cb
+```c
+MOCKABLE_FUNCTION(, SM_RESULT, sm_close_begin_with_cb, SM_HANDLE, sm, SM_OPENED_DRAINING_TO_CLOSE_COMPLETE_CALLBACK, callback, void*, callback_context);
+```
+
+`sm_close_begin_with_cb` asks from `sm` permission to exit `SM_OPENED` state (or one of its derived state) and return to `SM_CREATED`. `sm_close_begin_with_cb` invokes the `callback` function with `callback_context` before waiting for pending calls to become 0.
+
+**S_RS_SM_28_001: [** If `sm` is `NULL` then `sm_close_begin_with_cb` shall fail and return `SM_ERROR`. **]**
+
+**S_RS_SM_28_002: [** If `callback` is `NULL` then `sm_close_begin_with_cb` shall fail and return `SM_ERROR`. **]**
+
+**S_RS_SM_28_003: [** `sm_close_begin_with_cb` shall behave as if `sm_close_begin` was called except as follows.  **]**
```

Comment:
>shall behave as if `sm_close_begin` was called [](http://example.com/codeflow?start=47&length=46)

I would respec both sm_close_being and sm_close_begin_with_cb to call some other internal function (let's say "sm_close_begin_impl"). And that function has 1 liner that says "if CB!=NULL then CB(context)".

sm_close_begin can pass a NULL cb, sm_close_being_with_cb would pass the non-NULL cb.

The above helps with requirements (and also mocks).

Reply:
Seems we have sm_close_begin_internal, so spec that instead.

Reply:
I was with the understanding that we do not add specs for internal functions/implementations. Do you want me to add separate specs for sm_close_begin_internal?